internal/rbd/replicationcontrollerserver.go

```diff
@@ -321,31 +321,56 @@ func (rs *ReplicationServer) DisableVolumeReplication(ctx context.Context,
 	case librbd.MirrorImageDisabling:
 		return nil, status.Errorf(codes.Aborted, "%s is in disabling state", volumeID)
 	case librbd.MirrorImageEnabled:
-		if !force && !mirroringInfo.Primary {
-			return nil, status.Error(codes.InvalidArgument, "image is in non-primary state")
-		}
-		err = rbdVol.disableImageMirroring(force)
-		if err != nil {
-			util.ErrorLog(ctx, err.Error())
+		return disableVolumeReplication(rbdVol, mirroringInfo, force)
+	default:
+		// TODO: use string instead of int for returning valid error message
+		return nil, status.Errorf(codes.InvalidArgument, "image is in %d Mode", mirroringInfo.State)
+	}
 
-			return nil, status.Error(codes.Internal, err.Error())
-		}
-		// the image state can be still disabling once we disable the mirroring
-		// check the mirroring is disabled or not
-		mirroringInfo, err = rbdVol.getImageMirroringInfo()
-		if err != nil {
-			util.ErrorLog(ctx, err.Error())
+	return &replication.DisableVolumeReplicationResponse{}, nil
+}
 
-			return nil, status.Error(codes.Internal, err.Error())
+func disableVolumeReplication(rbdVol *rbdVolume,
```

Comment:
Is there a reason why this is not a function of `rbdVolume`? You could have the `mirroringInfo` argument as part of the `rbdVolume` struct, and drop the parameter here.

Reply:
Made `getLocalState` as the methods of the `rbdVolume` but didn't made `disableVolumeReplication`  a method of `rbdVolume` as this function returns the appropriate GRPC error codes and IMO the core function which is the methods of rbdVolume should not deal with GRPC methods. and as this is an extension to core rbd functions didn't add `mirroringInfo` as the member of the `rbdVolume` structure.

Am planning to rework core files with interfaces and breaking major functions into smaller helper functions to write more unit testings. I will work on creating a new structure that embeds the `rbdImage` core structure and moves these helper functions as an interface which will help us to write more unit testing.